### PR TITLE
ci(jenkins): support different targets

### PR DESCRIPTION
**Pull request by v1v**
## What?

Simplify the build and test stages by using Docker as much as possible or be able to configure the required tools in the CI workers.

## Why?

Simplify the requirements to run the builds, so, docker is the one in charge to drive what SDK is required.

image

## Tasks
- [x] Cache docker images in the Workers. 
- [x] Linux stages
- [x] Windows dotnet stages
- [x] Windows msbuild stages. Requirements [here](https://docs.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-build-tools?view=vs-2019)
- [x] Enable dotnet for windows - msbuild. Beta ONLY 

### MSBuild dependencies
- `includeOptionals` is required to get rid of the below error

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.CSharp.CurrentVersion.targets(324,5): 
error MSB4019: The imported project "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn\Microsoft.CSharp.Core.targets" was not found. 
Confirm that the expression in the Import declaration "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn\Microsoft.CSharp.Core.targets" is correct, and that the file exists on disk. 
[C:\Users\jenkins.PACKER-5DA3F2D9\workspace\tnet_apm-agent-dotnet-mbp_PR-646\a\apm-agent-dotnet\test\Elastic.Apm.EntityFrameworkCore.Tests\Elastic.Apm.EntityFrameworkCore.Tests.csproj]
```
- `visualstudio2019-workload-webbuildtools` is not required for the MSBuild.

- `.ci/windows/prepare-test.bat` was created to run the tests in the MSBuild

image

otherwise, there will be an error

```
 C:\Users\jenkins\workspace\tnet_apm-agent-dotnet-mbp_PR-646\apm-agent-dotnet\sample\AspNetFullFrameworkSampleApp\AspNetFullFrameworkSampleApp.csproj(263,3): error MSB4019: The imported project "C:\Program Files\dotnet\sdk\3.1.100\Microsoft\VisualStudio\v16.0\WebApplications\Microsoft.WebApplication.targets" was not found. Confirm that the expression in the Import declaration "C:\Program Files\dotnet\sdk\3.1.100\Microsoft\VisualStudio\v16.0\WebApplications\Microsoft.WebApplication.targets" is correct, and that the file exists on 
```

## Follow-ups
-  Global.json seems to be the one driving the SDK validation, so does it make sense to read the file and grab the version to be used when configuring the OS containers to run the builds?

- As soon as our infra provides the VS2019 installation then we can get rid of `.ci/windows/msbuild-tools.ps1` and therefore the builds will be 15 minutes faster or so

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -184,6 +184,7 @@ pipeline {
                     }
                     post {
                       always {
+                        archiveArtifacts(allowEmptyArchive: true, artifacts: "${BASE_DIR}/target/diag.log,${BASE_DIR}/target/TestResults.xml")
```
Comment (v1v) [inline Jenkinsfile]: This is only to help with debugging

File: Jenkinsfile [hunk 2]
```diff
@@ -288,6 +289,7 @@ pipeline {
...
```
Comment (v1v) [inline Jenkinsfile]: This is only to help with debugging

File: Jenkinsfile [hunk 3]
```diff
@@ -390,12 +392,10 @@ def cleanDir(path){
...
 
 def dotnet(Closure body){
-  def home = "/tmp"
-  def dotnetRoot = "/${home}/.dotnet"
-  def path = "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/${home}/bin:${dotnetRoot}:${dotnetRoot}/bin:${dotnetRoot}/tools"
-  docker.image('mcr.microsoft.com/dotnet/core/sdk:2.1.505').inside("-e HOME='${home}' -e PATH='${path}'"){
-    sh 'chmod +x .ci/linux/tools.sh'
-    sh '.ci/linux/tools.sh'
+  def home = '/tmp'
+  def dotnetRoot = '/usr/share/dotnet'
```
Comment (v1v) [inline Jenkinsfile]: This is the location of the dotNet installation, although worth to check whether it's required when using `mcr.microsoft.com/dotnet/core/sdk:3.1.100`

File: .ci/windows/tools.ps1 [hunk 4]
```diff
@@ -10,7 +10,6 @@ Add-WindowsFeature Web-Asp-Net45 ;
 
 # Install .Net SDKs
 & choco install dotnetcore-sdk -m -y --no-progress -r --version 3.1.100
-& choco install dotnetcore-sdk -m -y --no-progress -r --version 2.1.505
```
Comment (v1v) [inline .ci/windows/tools.ps1]: Unrequired if 3.1.100 is the one required

**Comment by gregkalapos** [comment]
🎉 🚀  Awesome! Thank you so much @v1v!

**Pull request merged**